src/Components/Blazor/WebAssembly.Authentication/src/Interop/AuthenticationService.ts

```diff
@@ -0,0 +1,319 @@
+                }
+            };
+        } else {
+            try {
+                const parameters = request && request.scopes ?
+                    { scope: request.scopes.join(' ') } : undefined;
+
+                const newUser = await this._userManager.signinSilent(parameters);
+
+                return {
+                    status: AccessTokenResultStatus.Success,
+                    token: {
+                        grantedScopes: newUser.scopes,
+                        expires: getExpiration(newUser.expires_in),
+                        value: newUser.access_token
+                    }
+                };
+
+            } catch (e) {
+                return {
+                    status: AccessTokenResultStatus.RequiresRedirect
+                };
+            }
+        }
+
+        function hasValidAccessToken(user: User | null): user is User {
+            return !!(user && user.access_token && !user.expired && user.scopes);
+        }
+
+        function getExpiration(expiresIn: number) {
+            const now = new Date();
+            now.setTime(now.getTime() + expiresIn * 1000);
+            return now;
+        }
+
+        function hasAllScopes(request: AccessTokenRequestOptions | undefined, currentScopes: string[]) {
+            const set = new Set(currentScopes);
+            if (request && request.scopes) {
+                for (let current of request.scopes) {
+                    if (!set.has(current)) {
+                        return false;
+                    }
+                }
+            }
+
+            return true;
+        }
+    }
+
+    async signIn(state: any) {
+        try {
+            await this._userManager.clearStaleState();
+            const silentUser = await this._userManager.signinSilent(this.createArguments());
+            return this.success(state);
+        } catch (silentError) {
+            await this._userManager.clearStaleState();
+            // User might not be authenticated, fallback to popup authentication
+            console.log("Silent authentication error: ", silentError);
+
+            try {
+                await this._userManager.signinRedirect(this.createArguments(state));
+                return this.redirect();
+            } catch (redirectError) {
+                console.log("Redirect authentication error: ", redirectError);
+                return this.error(redirectError);
+            }
+        }
+    }
+
+    async completeSignIn(url: string) {
+        const requiresLogin = await this.loginRequired(url);
+        const stateExists = await this.stateExists(url);
+        try {
+            const user = await this._userManager.signinCallback(url);
+            if (window.self !== window.top) {
+                return this.operationCompleted();
+            } else {
+                return this.success(user && user.state);
+            }
+        } catch (error) {
+            if (requiresLogin || window.self !== window.top || !stateExists) {
```

Comment:
This is too subtle for me to be confident I'm able to think through all the implications here.

Does this code come from the other SPA templates? Has it already been through some security review?

Reply:
There are no security concerns in this bits, unless you are thinking of something specific. In any case, I've filed to deal with this as part of https://github.com/dotnet/aspnetcore/issues/19053.